subsys/usb/class/mass_storage.c

```diff
@@ -491,6 +491,11 @@ static bool infoTransfer(void)
 
 static void fail(void)
 {
+	if (cbw.DataLength) {
+		/* Stall data stage */
+		usb_ep_set_stall(mass_ep_data[MSD_IN_EP_IDX].ep_addr);
+	}
+
```

Comment:
Why this approach now? That was literally the case You mentioned before, endpoint is stalled and there is attempt to send the data over stalled endpoint.

Reply:
Because it follows the exiting practice in the code, see also https://github.com/zephyrproject-rtos/zephyr/pull/23191#discussion_r388332142, and did not found a detailed definition of endpoint state after clear halt. I tested it for one day before I opened PR. I am also still working on mass_storage.c and will open an other cleanup PR.